File: ios/Sources/MoQKit/Codecs/VideoFrameProcessor.swift
import MoQKitFFI
    import CoreMedia
    import Foundation

    /// Resolves video format descriptions and preprocesses payloads for AVSampleBufferDisplayLayer.
    ///
    /// Strategy is determined by whether a decoder configuration record is present:
    /// - **Has description** (e.g. avc1, hev1, hvc1): payloads arrive length-prefixed (AVCC/HVCC)
    ///   and are passed through as-is — AVSampleBufferDisplayLayer expects this format.
    /// - **No description** (e.g. avc3, hev3): payloads are Annex B. Parameter sets are extracted
    ///   in-band from the first keyframe; subsequent frames are converted to length-prefixed.
    enum VideoCodec {
        case h264
        case hevc
        case av1

        init?(string: String) {
            let s = string.lowercased()
            if s.hasPrefix("avc") { self = .h264 }
            else if s.hasPrefix("hev") || s.hasPrefix("hvc") { self = .hevc }
            else if s.hasPrefix("av0") { self = .av1 }
            else { return nil }
        }
    }

    final class VideoFrameProcessor: @unchecked Sendable {
        private let hasDescription: Bool
        private let codec: VideoCodec
        private let codedWidth: Int32
        private let codedHeight: Int32
        private var formatDescription: CMFormatDescription?
        private let lock = NSLock()

        init(config: MoqVideo) throws {
            guard let codec = VideoCodec(string: config.codec) else {
                throw MoQSessionError.unsupportedCodec(config.codec)
            }
            self.hasDescription = config.description != nil
            self.codec = codec
            self.codedWidth = Int32(config.coded?.width ?? 0)
            self.codedHeight = Int32(config.coded?.height ?? 0)

            if hasDescription {
                self.formatDescription = try SampleBufferFactory.makeVideoFormatDescription(
                    from: config)
                MoQLogger.player.debug(
                    "VideoFrameProcessor: format description ready for codec=\(config.codec)")
            } else {
                MoQLogger.player.debug(
                    "VideoFrameProcessor: no description for codec=\(config.codec) — deferring to in-band parameter sets"
                )
            }
        }

        /// Whether this processor has a format description ready.
        var hasFormatDescription: Bool {
            lock.lock()
            defer { lock.unlock() }
            return formatDescription != nil
        }

        /// Whether this processor can eventually produce frames.
        var canProcess: Bool {
            return hasFormatDescription || !hasDescription
        }

        /// Process a raw frame payload into a CMSampleBuffer.
        ///
        /// Returns nil if the format description isn't available yet
        /// (waiting for the first in-band keyframe).
        func process(payload: Data, timestampUs: UInt64, keyframe: Bool)
            throws -> CMSampleBuffer?
        {
            lock.lock()

            if !hasDescription && formatDescription == nil {
                // Annex B path: extract parameter sets from the first keyframe
                guard keyframe else {
                    lock.unlock()
                    MoQLogger.player.debug(
                        "Dropping non-keyframe: waiting for in-band parameter sets")
                    return nil
                }

                do {
                    guard let fmt = try extractInBandFormatDescription(from: payload) else {
                        lock.unlock()
                        return nil
                    }
                    formatDescription = fmt
                } catch {
                    lock.unlock()
                    MoQLogger.player.error(
                        "Failed to build format description from in-band params: \(error)")
                    return nil
                }
            }

            guard let fmt = formatDescription else {
                lock.unlock()
                MoQLogger.player.debug("Waiting for format description, skipping frame")
                return nil
            }

            lock.unlock()

            // AV1 OBU temporal units are already in the correct format for
            // AVSampleBufferDisplayLayer and never need Annex B → length-prefix conversion.
            // Has-description H.264/H.265 payloads are already length-prefixed — pass through.
            // Only in-band H.264/H.265 (no out-of-band description) require the conversion.
            let processedPayload =
                (hasDescription || codec == .av1) ? payload : AnnexBDemuxer.toLengthPrefixed(payload)

            return try SampleBufferFactory.makeSampleBuffer(
                payload: processedPayload, timestampUs: timestampUs,
                formatDescription: fmt)
        }

        // MARK: - Private

        private func extractInBandFormatDescription(from payload: Data) throws
            -> CMFormatDescription?
        {
            switch codec {
            case .av1:
                guard let seqHeader = AV1Utils.extractSequenceHeader(from: payload) else {
                    MoQLogger.player.debug("Keyframe lacks AV1 sequence header, dropping")
                    return nil
                }
                MoQLogger.player.debug(
                    "Extracted in-band AV1 sequence header: \(seqHeader.count) bytes")
                return try SampleBufferFactory.makeAV1FormatDescriptionFromSequenceHeader(
                    seqHeader, width: codedWidth, height: codedHeight)
            case .hevc:
                guard let params = H265Utils.extractParameterSets(from: payload) else {
                    MoQLogger.player.debug("Keyframe lacks H.265 VPS/SPS/PPS, dropping")
                    return nil
                }
                MoQLogger.player.debug(
                    "Extracted in-band HEVC parameter sets: vps=\(params.vps.count) sps=\(params.sps.count) pps=\(params.pps.count)"
                )
                return try SampleBufferFactory.makeHEVCFormatDescriptionFromParameterSets(
                    vps: params.vps, sps: params.sps, pps: params.pps)
            case .h264:
                guard let params = H264Utils.extractParameterSets(from: payload) else {
                    MoQLogger.player.debug("Keyframe lacks H.264 SPS/PPS, dropping")
                    return nil
                }
                MoQLogger.player.debug(
                    "Extracted in-band H.264 parameter sets: sps=\(params.sps.count) pps=\(params.pps.count)"
                )
                return try SampleBufferFactory.makeH264FormatDescriptionFromParameterSets(
                    sps: params.sps, pps: params.pps)
            }
        }
    }

File: android/moqkit/moqkit/src/main/java/com/swmansion/moqkit/VideoFrameProcessor.kt
package com.swmansion.moqkit

import android.media.MediaFormat
import android.util.Log
import uniffi.moq.MoqVideo
import java.nio.ByteBuffer

private const val TAG = "VideoFrameProcessor"

/**
 * Selects the right payload transform and CSD extraction strategy based on the video config.
 *
 * - **Has description** (avc1/hev1/hvc1): payloads are AVCC/HVCC length-prefixed → convert to
 *   Annex B via [avccToAnnexB]; CSD parsed from the config record by [MediaFactory].
 * - **No description** (avc3/hev3): payloads are already Annex B → passthrough; CSD extracted
 *   in-band from the first keyframe containing parameter sets.
 *
 * Consumers call [processPayload] and get back ready-to-decode Annex B bytes.
 */
internal class VideoFrameProcessor(private val config: MoqVideo) {

    private val transform: (ByteArray) -> ByteArray
    private var format: MediaFormat? = null

    /** True once a MediaFormat with CSD is available and the decoder can be configured. */
    val isReady: Boolean get() = format != null

    init {
        transform = if (config.description != null) { payload -> payload.prefixLengthToAnnexB() }
                     else { payload -> payload }

        if (config.description != null) {
            format = MediaFactory.makeVideoFormat(config)
            if (format != null) {
                Log.d(TAG, "Format ready immediately: $format")
            } else {
                Log.w(TAG, "makeVideoFormat returned null for codec=${config.codec}")
            }
        } else {
            Log.d(TAG, "No description — deferring CSD extraction for codec=${config.codec}")
        }
    }

    /** Returns the MediaFormat once CSD is available, null otherwise. */
    fun getFormat(): MediaFormat? = format

    /**
     * Process a compressed video frame payload.
     *
     * @return Annex B bytes ready for MediaCodec, or null if the frame should be dropped
     *         (e.g., waiting for a keyframe with parameter sets).
     */
    fun processPayload(payload: ByteArray, keyframe: Boolean): ByteArray? {
        if (!isReady) {
            Log.i(TAG, "Video processor not ready")
            if (!keyframe) {
                Log.w(TAG, "Expected a keyframe when in not-ready state")
                return null
            }

            val mime = MediaFactory.videoMime(config.codec)
            if (mime == null) {
                Log.w(TAG, "Unsupported codec for in-band CSD: ${config.codec}")
                return null
            }

            val width = config.coded?.width?.toInt() ?: 1920
            val height = config.coded?.height?.toInt() ?: 1080
            val fmt = MediaFormat.createVideoFormat(mime, width, height)

            when (mime) {
                MediaFormat.MIMETYPE_VIDEO_AVC -> {
                    val params = AnnexBUtils.extractH264ParameterSets(payload)
                    if (params == null) {
                        Log.d(TAG, "Keyframe lacks H.264 SPS/PPS, dropping")
                        return null
                    }
                    Log.d(TAG, "Extracted in-band SPS (${params.sps.size}B) + PPS (${params.pps.size}B)")
                    fmt.setByteBuffer("csd-0", ByteBuffer.wrap(params.sps))
                    fmt.setByteBuffer("csd-1", ByteBuffer.wrap(params.pps))
                }
                MediaFormat.MIMETYPE_VIDEO_HEVC -> {
                    val csd = AnnexBUtils.extractH265ParameterSets(payload)
                    if (csd == null) {
                        Log.d(TAG, "Keyframe lacks H.265 VPS/SPS/PPS, dropping")
                        return null
                    }
                    Log.d(TAG, "Extracted in-band HEVC parameter sets (${csd.size}B)")
                    fmt.setByteBuffer("csd-0", ByteBuffer.wrap(csd))
                }
                else -> {
                    Log.e(TAG, "Unknown mime type $mime")
                    return null
                }
            }

            format = fmt
            Log.d(TAG, "Format now ready: $fmt")
        }

        return transform(payload)
    }
}
